File: 프로그래머스/3/77886. 110 옮기기/110 옮기기.py
def ext(x):
    stk = [];
    num1 = 0
    cnt = 0
    for c in x:
        if c == "1":
            stk.append(c);
            num1 += 1
        else:
            if num1 >= 2:
                stk.pop();
                stk.pop();
                num1 -= 2;
                cnt += 1
            else:
                stk.append(c)
                num1 = 0
    
    return ''.join(stk), cnt

def ins(x, cnt):
    idx_111 = x.find("111")
    if idx_111 != -1:
        return x[:idx_111] + "110" * cnt + x[idx_111:]
    else:
        idx_last0 = x.rfind("0")
        if idx_last0 != -1:
            return x[:idx_last0 + 1] + "110" * cnt + x[idx_last0 + 1:]
        else:
            return "110" * cnt + x

def solution(s):
    answer = []
    for x in s:
        new_x, cnt = ext(x)
        answer.append(ins(new_x, cnt))
    return answer

File: 프로그래머스/3/77886. 110 옮기기/110 옮기기.cpp
#include <string>
#include <vector>
#include <stack>
#include <algorithm>

using namespace std;

vector<string> solution(vector<string> s) {
    vector<string> answer;
    for(string x : s){
        // extract 110
        stack<char> stk;
        int num1 = 0; // 연속해서 1이 나온 개수
        int cnt = 0; // 110 개수
        for(char c : x){
            if(c == '1'){
                stk.push(c);
                num1++;
            }
            else{
                if(num1 >= 2){
                    stk.pop();
                    stk.pop();
                    num1 -= 2;
                    cnt++;
                }
                else{
                    stk.push(c);
                    num1 = 0;
                }
            }
        }
        string new_x = "";
        while(!stk.empty()){
            new_x += stk.top();
            stk.pop();
        }
        reverse(new_x.begin(), new_x.end());
        
        // insert 110
        string to_insert = "";
        for(int i = 0; i < cnt; i++) to_insert += "110";
        
        int idx_111 = new_x.find("111");
        if(idx_111 != string::npos){
            new_x.insert(idx_111, to_insert);
        }
        else{
            int idx_last0 = new_x.rfind('0');
            if(idx_last0 != string::npos) new_x.insert(idx_last0 + 1, to_insert);
            else new_x = to_insert + new_x;
        }
        
        // answer
        answer.push_back(new_x);
    }
    return answer;
}